### PR TITLE
[riverpod_generator] apply meta annotations on generated providers

### PR DESCRIPTION
**Pull request by SunlightBro**
#2886

Added function `metaAnnotations` to riverpod_generator, that gets used in
ClassBasedProviderTemplate, FamilyTemplate, FunctionalProviderTemplate

### CODE REVIEW

File: packages/riverpod_generator/lib/src/riverpod_generator.dart [hunk 1]
```diff
@@ -20,6 +20,28 @@ String providerDocFor(Element element) {
       : '${element.documentationComment}\n///\n/// Copied from [${element.name}].';
 }
 
+final metaAnnotationsTypeChecker = TypeChecker.any([
+  const TypeChecker.fromRuntime(Deprecated),
+  TypeChecker.fromRuntime(experimental.runtimeType),
```
Comment (SunlightBro) [inline packages/riverpod_generator/lib/src/riverpod_generator.dart:25]: The classes of annotations from meta are all private.
```dart
const _Experimental experimental = _Experimental();

class _Experimental {
  const _Experimental();
}
```
Is there a better way to get it's type?

File: packages/riverpod_generator/test/annotated_test.dart [hunk 2]
```diff
@@ -0,0 +1,11 @@
+import 'package:riverpod/riverpod.dart';
+import 'package:test/test.dart';
+
+import 'integration/annotated.dart';
+
+void main() {
+  test('Applied meta annotations to provider', () {
+    //TODO (SunlightBro): check if annotations are correctly applied.
+    expect(functionalProvider, isA<AutoDisposeProvider<String>>());
```
Comment (SunlightBro) [inline packages/riverpod_generator/test/annotated_test.dart:9]: Have to check if there is a Matcher for checking annotations in test, or write my own.
Or do you have a better way to test for annotations?

File: packages/riverpod_generator/lib/src/templates/family.dart [hunk 3]
```diff
@@ -229,6 +230,11 @@ ${parameters.map((e) => '        ${e.name}: ${e.name},\n').join()}
     });
 
     final docs = providerDocFor(provider.providerElement.element);
+    var meta = '';
+    final node = provider.node;
+    if (node is AnnotatedNode) {
```
Comment (SunlightBro) [inline packages/riverpod_generator/lib/src/templates/family.dart:235]: Both `ClassDeclaration` and `FunctionDeclaration` are of type `AnnotatedNode`, so theoretically we could write
```dart
final meta = metaAnnotations((provider.node as AnnotatedNode).metadata);
```

File: packages/riverpod_generator/test/annotated_test.dart (hunk 2 above)
Comment (rrousselGit) [reply packages/riverpod_generator/test/annotated_test.dart:10]: Use parseString/file or resolveFile from analyzer

File: packages/riverpod_generator/lib/src/templates/family.dart (hunk 3 above)
Comment (rrousselGit) [reply packages/riverpod_generator/lib/src/templates/family.dart:235]: An "is" check shouldn't be needed then.
Consider updating riverpod_analizer_utils to have "node" be an AnnotatedNode

Comment (rrousselGit) [reply packages/riverpod_generator/lib/src/templates/family.dart:235]: You'd then remove this if condition

**Comment by SunlightBro** [comment]
Will also add 
`deprecated_member_use_from_same_package` to RiverpodGenerator's ignore_for_file.

File: packages/riverpod_generator/lib/src/riverpod_generator.dart (hunk 1 above)
Comment (rrousselGit) [reply packages/riverpod_generator/lib/src/riverpod_generator.dart:25]: Don't use TypeChecker to begin with.

Meta annotations are official. There's straight up a `isExperimental`/`isDeprecated` getter in the analyzer

Comment (rrousselGit) [reply packages/riverpod_generator/lib/src/riverpod_generator.dart:25]: Actually isExperimental might be the exception. It seems to be missing. We probably don't care too much about it anyway. I doubt people use it

**Pull request merged**